ghcide/src/Text/Fuzzy/Parallel.hs

```diff
@@ -1,96 +1,90 @@
+      --    is too much of performance penalty for fuzzy search. Again, anyway HLS does not
+      --    attempt to do justice to Unicode: proper Unicode text matching requires
+      --    `unicode-transforms` and friends.
+      -- Altogether we sacrifice correctness for the sake of performance, which
+      -- is a right trade-off for fuzzy search.
+      | pByte <- TA.unsafeIndex pArr currPOff
+      , sByte <- TA.unsafeIndex sArr currSOff
+      -- First byte (currPOff == pOff) should match exactly, otherwise - up to case.
+      , pByte == sByte || (currPOff /= pOff && pByte == toLowerAscii sByte)
+      = let curr = currScore * 2 + 1 in
+        go (totalScore + curr) curr (currPOff + 1) (currSOff + 1)
+      | otherwise
+      = go totalScore 0 currPOff (currSOff + 1)
+
+    toLowerAscii w = if (w - 65) < 26 then w .|. 0x20 else w
 
 -- | The function to filter a list of values by fuzzy search on the text extracted from them.
-filter :: (TextualMonoid s)
-       => Int      -- ^ Chunk size. 1000 works well.
-       -> Int      -- ^ Max. number of results wanted
-       -> s        -- ^ Pattern.
-       -> [t]      -- ^ The list of values containing the text to search in.
-       -> s        -- ^ The text to add before each match.
-       -> s        -- ^ The text to add after each match.
-       -> (t -> s) -- ^ The function to extract the text from the container.
-       -> [Scored t] -- ^ The list of results, sorted, highest score first.
-filter chunkSize maxRes pattern ts pre post extract = runST $ do
-  let v = V.mapMaybe id
-             (V.map (\t -> match pattern' t pre post extract) (V.fromList ts)
-             `using`
-             parVectorChunk chunkSize (evalTraversable forceScore))
-      perfectScore = score $ fromMaybe (error $ T.toString undefined pattern) $
-        match pattern' pattern' "" "" id
-  return $ partialSortByAscScore maxRes perfectScore v
+filter :: Int           -- ^ Chunk size. 1000 works well.
+       -> Int           -- ^ Max. number of results wanted
+       -> T.Text        -- ^ Pattern.
+       -> [t]           -- ^ The list of values containing the text to search in.
+       -> (t -> T.Text) -- ^ The function to extract the text from the container.
+       -> [Scored t]    -- ^ The list of results, sorted, highest score first.
+filter chunkSize maxRes pattern ts extract = partialSortByAscScore maxRes perfectScore (concat vss)
   where
       -- Preserve case for the first character, make all others lowercase
-      pattern' = case T.splitCharacterPrefix pattern of
-          Just (c, rest) -> T.singleton c <> T.map toLower rest
-          _              -> pattern
+      pattern' = case T.uncons pattern of
+        Just (c, rest) -> T.cons c (T.toLower rest)
+        _              -> pattern
+      vss = map (mapMaybe (\t -> flip Scored t <$> match pattern' (extract t))) (chunkList chunkSize ts)
+        `using` parList (evalList rseq)
```

Comment:
Please simplify with `parListChunk`

Reply:
What exactly do you have in mind? Something like 
```haskell
      vs = mapMaybe (\t -> flip Scored t <$> match pattern' (extract t)) ts
        `using` parListChunk chunkSize rseq
```
does not make much sense, because it chunks outputs instead of inputs. Alternatively,
```haskell
      vs = catMaybes (map (\t -> flip Scored t <$> match pattern' (extract t)) ts
        `using` parListChunk chunkSize rseq)
```
could end up allocating `Maybe` thunks, because parallel computation would be a barrier for fusion with `catMaybes`.